server/lib/pbench_server/pbench_server.py

```diff
@@ -0,0 +1,130 @@
+        return response
+
+
+class Check(Resource):
+    def get(self):
+        result = False
+
+        if not request.headers.get("filename"):
+            logger.debug("Missed filename in header")
+            abort(400, message='Missing filename header in request')
+        filename = request.headers.get("filename")
+        tarball_md5 = f"{filename}.md5.check"
+
+        if not os.path.exists(filename):
+            logger.debug("Filename %s does not exist", filename)
+            abort(400, message='Filename does not appear to be a valid one')
+        if not os.path.exists(tarball_md5):
+            logger.debug("MD5 check sum does not exist for %s", filename)
+            abort(400, message='Missing MD5 check sum')
+
+        try:
+            hash_md5 = hashlib.md5()
+            with open(filename, "rb") as tar:
+                for chunk in iter(lambda: tar.read(4096), b""):
+                    hash_md5.update(chunk)
+
+                with open(tarball_md5, "rb") as check_sum:
+                    hash_check = check_sum.read()
+                    if hash_md5.hexdigest() == hash_check.decode("utf-8"):
+                        result = True
+        except Exception:
+            logger.error("md5sum failed for %s, skipping", filename)
+            if os.path.exists(filename):
+                os.remove(filename)
+            if os.path.exists(tarball_md5):
+                os.remove(tarball_md5)
+            abort(400, message=f"md5sum check failed for {filename}")
+
+        if result:
+            response = jsonify(dict(result='md5sum check is correct'))
+            response.status_code = 200
+            return response
+        else:
+            abort(400, message=f"md5sum check failed for {filename}")
+
+
+class Upload(Resource):
+    def post(self):
+        if not request.headers.get("filename"):
+            logger.debug("Missed filename in header")
+            abort(400, message='Missing filename header in request')
+        filename = secure_filename(request.headers.get("filename"))
+        logger.debug(f"Receiving file: {filename}")
+        if not allowed_file(filename):
+            logger.debug("Bad file extension received")
+            abort(400, message='File extension not supported. Only .xz')
+        upload_directory = conf.get("pbench-receive-dir-prefix")
+        if not upload_directory:
+            logger.debug("Missing config variable for pbench-receive-dir-prefix")
+            abort(400, message="Missing config variable for pbench-receive-dir-prefix")
+        if not os.path.exists(upload_directory):
+            os.mkdir(upload_directory)
+        full_path = os.path.join(upload_directory, filename)
+        with open(full_path, "wb") as f:
+            chunk_size = 4096
+            logger.debug("Writing chunks")
+
+            while True:
+                chunk = request.stream.read(chunk_size)
+                if len(chunk) == 0:
+                    break
+
+                f.write(chunk)
```

Comment:
Why can't we calculate the MD5 check sum of the file as we are reading it off the wire, like https://github.com/distributed-system-analysis/pbench/blob/f32e5b4eeae9f99789ecc824b7bfca6aed509535/agent/util-scripts/pbench-tool-data-sink#L65?

Then we could get rid of the `/check` end point.